src/protocol/templates/ping_protocol.json

```diff
@@ -1,212 +1,511 @@
-                  },
-                  {
-                      "name":"confidence",
-                      "type":"u8"
-                  },
-                  {
-                      "name":"pulse_usec",
-                      "type":"u16"
-                  },
-                  {
-                      "name":"ping_number",
-                      "type":"u32"
-                  },
-                  {
-                      "name":"start_mm",
-                      "type":"u32"
-                  },
-                  {
-                      "name":"length_mm",
-                      "type":"u32"
-                  },
-                  {
-                      "name":"gain_index",
-                      "type":"u32"
-                  },
-                  {
-                      "name":"num_points",
-                      "type":"u16"
-                  },
-                  {
-                      "name":"data",
-                      "type":"u8[200]"
-                  }
-              ]
-          },
-          "range":{
-              "id":"1110",
-              "payload":[
-                  {
-                      "name":"start_mm",
-                      "type":"u32"
-                  },
-                  {
-                      "name":"length_mm",
-                      "type":"u32"
-                  }
-              ]
-          },
-          "mode":{
-              "id":"1111",
-              "payload":[
-                  {
-                      "name":"auto_manual",
-                      "type":"u8"
-                  }
-              ]
-          },
-          "rate":{
-              "id":"1112",
-              "payload":[
-                  {
-                      "name":"msec_per_ping",
-                      "type":"u16"
-                  }
-              ]
-          },
-          "gain":{
-              "id":"1113",
-              "payload":[
-                  {
-                      "name":"gain_index",
-                      "type":"u8"
-                  }
-              ]
-          },
-          "pulse":{
-              "id":"1114",
-              "payload":[
-                  {
-                      "name":"pulse_usec",
-                      "type":"u16"
-                  }
-              ]
+  "requirements": {
+    "discovery": "discovery of connected devices, types, and firmwares",
+    "packed fields": "memory alignment"
+  },
+  "format": [
+    {
+      "name": "start1",
+      "type": "u8",
+      "index": "0"
+    },
+    {
+      "name": "start2",
+      "type": "u8",
+      "index": "1"
+    },
+    {
+      "name": "length",
+      "type": "u16",
+      "index": "2-3"
+    },
+    {
+      "name": "message_id",
+      "type": "u16",
+      "index": "4-5"
+    },
+    {
+      "name": "src_device_id",
+      "type": "u8",
+      "index": "6"
+    },
+    {
+      "name": "dst_device_id",
+      "type": "u8",
+      "index": "7"
+    },
+    {
+      "name": "payload",
+      "type": "-",
```

Comment:
u8, because it will be a byte array

Reply:
The type inside of each payload is message specific. u8 is misleading imo.